### PR TITLE
Store RNA analysis in Housekeeper

### PR DESCRIPTION
**Pull request by barrystokman**
This PR enables RNA analyses to be stored in Housekeeper.

See the user story [here](https://github.com/Clinical-Genomics/housekeeper/issues/49).
Related PR [here](https://github.com/Clinical-Genomics/housekeeper/pull/51).

**How to prepare for test**:
- [x] install on hasta stage `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh rna-analysis-in-hk`
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-housekeeper-stage.sh rna-analysis-in-hk`

- [x] activate stage: `us`
- [x] remove existing housekeeper bundles (if any): `housekeeper delete bundle finequagga'
- [x] remove existing analysis (if any) from clinical-api-stage: https://clinical-api-stage.scilifelab.se/admin/analysis/?search=finequagga

**How to test**:
- [x] run following command: `cg store analysis /home/proj/stage/rare-disease/cases/finequagga/analysis/finequagga_config.yaml`

**Expected test outcome**:
- [x] Printed to terminal:
```
new bundle added: finequagga, version 2019-11-21                                                                                                                                              
created new bundle version dir: /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21                                                                                                    
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/finequagga_config.yaml -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/finequagga_config.yaml              
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/mip_2019-11-21T14:31:02.log -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/mip_2019-11-21T14:31:02.l
og                                                                                                                                                                                            
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/finequagga_qc_sample_info.yaml -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/finequagga_qc_sample_info.ya
ml                                                                                                                                                                                            
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/finequagga_comb.vcf -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/finequagga_comb.vcf              
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/multiqc_report.html -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/multiqc_report.html              
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/multiqc_data.json -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/multiqc_data.json                  
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/finequagga.fam -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/finequagga.fam                        
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/qcmetrics_file.stub -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/qcmetrics_file.stub              
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/finequagga_comb_vep.vcf -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/finequagga_comb_vep.vcf      
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/finequagga_vcol.yaml -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/finequagga_vcol.yaml            
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/ACC5963A1_lanes_1234_trim_star_sorted_merged_md_splitncigar_brecal_haptc_filtered_ase_bootstr.vcf -> /home/proj/sta
ge/housekeeper-bundles/finequagga/2019-11-21/ACC5963A1_lanes_1234_trim_star_sorted_merged_md_splitncigar_brecal_haptc_filtered_ase_bootstr.vcf                                                
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/foo_ACC5963A1_lanes_1234_trim_star_sorted_merged_md_splitncigar_brecal_haptc_filtered_ase_bootstr.vcf -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/foo_ACC5963A1_lanes_1234_trim_star_sorted_merged_md_splitncigar_brecal_haptc_filtered_ase_bootstr.vcf
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/ACC5963A1_lanes_1234_trim_star_sorted_merged_md_splitncigar_brecal_haptc_filtered_ase.csv -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/ACC5963A1_lanes_1234_trim_star_sorted_merged_md_splitncigar_brecal_haptc_filtered_ase.csv
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/ACC5963A1_lanes_1234_trim_star_sorted_merged_md_splitncigar_brecal.bam -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/ACC5963A1_lanes_1234_trim_star_sorted_merged_md_splitncigar_brecal.bam
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/ACC5963A1_lanes_1234_trim_star_sorted_merged_strg_gffcmp.gtf -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/ACC5963A1_lanes_1234_trim_star_sorted_merged_strg_gffcmp.gtf
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/ACC5963A1_lanes_1234_trim_star_sorted_merged_md_metric -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/ACC5963A1_lanes_1234_trim_star_sorted_merged_md_metric
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/quant.sf -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/quant.sf
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/star-fusion.fusion_predictions.tsv -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/star-fusion.fusion
_predictions.tsv
linked file: /home/proj/stage/rare-disease/cases/finequagga/analysis/files/ACC5963A1_lanes_1234_trim_star_sorted_merged_strg.gtf -> /home/proj/stage/housekeeper-bundles/finequagga/2019-11-21
/ACC5963A1_lanes_1234_trim_star_sorted_merged_strg.gtf
included files in Housekeeper
```
- [x] Analysis with pipeline 'MIP + RNA' is added to clinical-api-stage: https://clinical-api-stage.scilifelab.se/admin/analysis/?search=finequagga
- [x] files are added to '/home/proj/stage/housekeeper-bundles/finequagga/2019-11-21/'
- [x] run `housekeeper get -V finequagga`, this should show the files and the assigned tags.
- [x] Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by @patrikgrenfeldt 
- [x] tests executed by @jemten @barrystokman 
- [x] "Merge and deploy" approved by @patrikgrenfeldt 
Thanks for filling in who performed the code review and the test!

This is minor **version bump** because all changes are backwards compatible

### CODE REVIEW

**Comment by barrystokman** [comment]
@ingkebil @patrikgrenfeldt I'd like to summarize what we're trying to accomplish here:

1) This PR uses test fixtures (case_config.yaml and case_qc_sampleinfo.yaml) that trigger pylint/flake8 linting errors regarding line length. I want the linter to ignore these files because I want the test data to be as much production-like as possible.

2) Snapshot automatically generates files with the outcome of unit tests. These files are .py files (but they hold data rather than production code!) and are stored in directories called `snapshots` in the directories of the respective unit tests (for instance `/cg/tests/apps/mip/snapshots`). I want the linter to ignore these files too, because having to go into these files to fix line length defeats the purpose of automatic generation.

What I've tried:

1. Add `# pylint: disable-all` to the snapshots and yamls. This disables the pytest linting, but not the flake8 linting. It's also another manual step that should be avoided.  :negative_squared_cross_mark: 
1. Use black for linting the snapshot files: black cannot fix the line length in these files. :negative_squared_cross_mark:
1. Add `--ignore=snapshots` to the pylint arguments in .gitlint.yaml. Adding `--ignore=snapshots` to the pylint on the command line works, but Travis seems to ignore this :negative_squared_cross_mark:

Kenny, can you summarize your attempts as well?
Patrik, any ideas? Not burdened by any knowledge whatsoever, I assumed it would be quite easy to permanently exclude certain files (based on location, name and/or file type) from linting...

**Comment by ingkebil** [comment]
What I've tried and doesn't work:
- I've added a pylintrc file which is referred from the gitlint file. In
there I added a MASTER block to give the `--ignore=snapshots` param.
- I removed the msg formatting in the gitlint file, mostly because I didn't
understand the syntax. Doesn't work. Did reveal that the gitlint file is
used by Travis as gitlint errored on faulty syntax.

**Comment by barrystokman** [comment]
I'm going to make an issue out of this thing, maybe we can do something with it later.

For now let's stick to the solution Patrik suggested when I talk to him about it this morning: handle it pragmatically and just move forward :)

**Comment by barrystokman** [comment]
@patrikgrenfeldt @ingkebil This is now ready to be reviewed
@jemten If you're looking at this, can you also review the expected test outcome in the PR to make sure I got everything?

**Comment by barrystokman** [comment]
The linting errors in test data will (yamls and snapshots) not be solved.

File: .travis.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ python:
 install:
   - pip install --upgrade pip
   - pip install cython
-  - pip install -U -r requirements.txt pytest-mock .
+  - pip install -U -r requirements.txt pytest-mock snapshottest .
```
Comment (ingkebil) [inline .travis.yml]:
```suggestion
  - pip install -U -r requirements.txt .
```

and add `snapshottest` to the `requirements-dev.txt`

File: .gitlint.yaml [hunk 2]
```diff
@@ -92,6 +92,7 @@ pylint:
     - --rcfile={DEFAULT_CONFIGS}/pylintrc
     - --output-format=text
     - --max-line-length=100
+    - --ignore=snapshots
```
Comment (ingkebil) [inline .gitlint.yaml]:
```suggestion
```

Comment (ingkebil) [reply .gitlint.yaml]: doesn't seem to work anyway :(

File: .travis.yml (hunk 1 above)
Comment (barrystokman) [reply .travis.yml]: As discussed, let's leave it like it is. Also, `snapshottest` is already in `requirements-dev.txt`.

**Comment by barrystokman** [comment]
Tested on stage:

image

File: README.md [hunk 3]
```diff
@@ -327,6 +327,8 @@ There's one file for storing all constants like how priority levels are translat
 
 Another module `/exc.py` contains the custom Exception classes that are used across the package.
 
+Some unit tests make use of [snapshottest][snapshottest]. To update existing snapshots run `pytest --snapshot-update`
```
Comment (ingkebil) [inline README.md]: Excellent! :+1:

File: cg/apps/mip/files.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        'case': case,
+        'config_file_path': data['config_file_analysis'],
+        'version': data['mip_version'],
+        'pedigree_path': data['pedigree_minimal'],
+        'bcftools_merge': data['recipe']['bcftools_merge']['path'],
+        'multiqc_html': data['recipe']['multiqc'][f'{case}_html']['path'],
+        'multiqc_json': data['recipe']['multiqc'][f'{case}_json']['path'],
+        'qcmetrics_path': qcmetrics_parse if not isinstance(qcmetrics_parse, dict) else None,
+        'vep_path': data['recipe']['varianteffectpredictor']['path'],
+        'vp_stderr_file': vp_stderr_parse if not isinstance(vp_stderr_parse, dict) else None,
```
[the remark this answers is not in the document]
Comment (barrystokman) [reply cg/apps/mip/files.py]: You mean the vp_stderr_file?

File: cg/apps/mip/files.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    """Get all paths to files of a given type. Use this method if the exact filename is not
+    known beforehand.
+    Args:
+        sample_data (dict): YAML file containing qc sample info
+        path_key (str)    : Type of file paths to retrieve
+
+    Returns:
+        list: paths to all files of given type
+    """
+    paths = [path['path'] for path in sample_data['recipe'][path_key].values()]
```
Comment (barrystokman) [inline cg/apps/mip/files.py]:
```suggestion
    paths = [file_data['path'] for file_data in sample_data['recipe'][path_key].values()]
```

File: tests/apps/mip/test_apps_mip_files.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+           cg/cg/tests/snapshots/snap_test_apps_mip_files.py
+           files_raw (dict): dict of raw .yaml files
+    """
+    # GIVEN an RNA sample info file
+    sampleinfo_raw = files_raw['rna_sampleinfo']
+
+    # WHEN parsing the file
+    sampleinfo_data = files.parse_sampleinfo_rna(sampleinfo_raw)
+
+    # THEN the result should contains the correct data
```
Comment (barrystokman) [inline tests/apps/mip/test_apps_mip_files.py]:
```suggestion
    # THEN the result should contains the data as specified by the MIP team
```

File: tests/apps/tb/mip/test_get_files.py [hunk 7]
```diff
@@ -138,3 +138,89 @@ def test_get_files(files_data) -> dict:
...
+
+def test_rna_analysis_multiple_some_true(files_raw):
+    """
+    Args:
+    files_raw (dict): With dicts from files
+    """
+    # GIVEN an raw sample info file with more than one sample
+    rna_sampleinfo = files_raw['rna_sampleinfo']
+
+    # WHEN the analysis type is not 'wts' for all samples
```
Comment (barrystokman) [inline tests/apps/tb/mip/test_get_files.py]:
```suggestion
    # WHEN the analysis type is not 'wts' for all samples, but only some
```

File: tests/conftest.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
 
 import pytest
 import ruamel.yaml
+from cg.apps.mip import files as mip_rna_files_api
 from cg.store import Store
 from trailblazer.mip import files as mip_files_api
```
Comment (barrystokman) [inline tests/conftest.py]:
```suggestion
from trailblazer.mip import files as mip_dna_files_api
```

File: tests/meta/store/test_mip_rna_build_bundle.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    # GIVEN config data of a "sharp" RNA run (not dry run)
+    mip_rna_config = files_data['rna_config']
+
+    # GIVEN sampleinfo input from a finished RNA analysis
+    rna_sampleinfo = files_data['rna_sampleinfo']
+
+    mip_rna_file_data = mip_rna.get_files_rna(mip_rna_config, rna_sampleinfo)
+
+    # THEN the result should contain the correct data
+
```
Comment (barrystokman) [inline tests/meta/store/test_mip_rna_build_bundle.py]:
```suggestion
```

File: tests/meta/store/test_mip_rna_build_bundle.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    # GIVEN config data of a "sharp" RNA run (not dry run)
+    mip_rna_config = files_data['rna_config']
+
+    # GIVEN sampleinfo input from a finished RNA analysis
+    rna_sampleinfo = files_data['rna_sampleinfo']
+
+    mip_rna_bundle = mip_rna.build_bundle_rna(mip_rna_config, rna_sampleinfo)
+
+    # THEN the result should contain the correct data
+
```
Comment (barrystokman) [inline tests/meta/store/test_mip_rna_build_bundle.py]:
```suggestion
```

File: tests/meta/store/test_mip_rna_build_bundle.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    """
+    # GIVEN config data of a "sharp" RNA run (not dry run)
+    mip_rna_config = files_data['rna_config']
+
+    # GIVEN sampleinfo input from a finished RNA analysis
+    rna_sampleinfo = files_data['rna_sampleinfo']
+
+    mip_rna_bundle = mip_rna.build_bundle_rna(mip_rna_config, rna_sampleinfo)
+
+    # THEN the result should contain the correct data
```
Comment (barrystokman) [inline tests/meta/store/test_mip_rna_build_bundle.py]:
```suggestion
    # THEN the result should contain the data as specified by the MIP version
```

File: tests/meta/store/test_mip_rna_build_bundle.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    """
+    # GIVEN config data of a "sharp" RNA run (not dry run)
+    mip_rna_config = files_data['rna_config']
+
+    # GIVEN sampleinfo input from a finished RNA analysis
+    rna_sampleinfo = files_data['rna_sampleinfo']
+
+    mip_rna_file_data = mip_rna.get_files_rna(mip_rna_config, rna_sampleinfo)
+
+    # THEN the result should contain the correct data
```
Comment (barrystokman) [inline tests/meta/store/test_mip_rna_build_bundle.py]:
```suggestion
    # THEN the result should contain the data as specified by the MIP bundle
```

File: tests/meta/store/test_mip_rna_build_bundle.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    mip_rna_config = files_data['rna_config']
+
+    # GIVEN sampleinfo input from a finished RNA analysis and a vp_srderr_file is present
+    rna_sampleinfo = files_data['rna_sampleinfo']
+    rna_sampleinfo['vp_stderr_file'] = '/some/path/file.name'
+
+    mip_rna_bundle = mip_rna.build_bundle_rna(mip_rna_config, rna_sampleinfo)
+
+    # THEN the result should contain the correct data
+
```
Comment (barrystokman) [inline tests/meta/store/test_mip_rna_build_bundle.py]:
```suggestion
```

File: tests/meta/store/test_mip_rna_build_bundle.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    # GIVEN config data of a "sharp" RNA run (not dry run)
+    mip_rna_config = files_data['rna_config']
+
+    # GIVEN sampleinfo input from a finished RNA analysis and a vp_srderr_file is present
+    rna_sampleinfo = files_data['rna_sampleinfo']
+    rna_sampleinfo['vp_stderr_file'] = '/some/path/file.name'
+
+    mip_rna_bundle = mip_rna.build_bundle_rna(mip_rna_config, rna_sampleinfo)
+
+    # THEN the result should contain the correct data
```
Comment (barrystokman) [inline tests/meta/store/test_mip_rna_build_bundle.py]:
```suggestion
    # THEN the result should contain the data as specified by the MIP bundle
```

File: cg/apps/mip/files.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jemten) [reply cg/apps/mip/files.py]: yes
